consensus/core/src/sign.rs

```diff
@@ -17,6 +17,45 @@ pub enum Error {
 
     #[error("Secp256k1 -> {0}")]
     Secp256k1Error(#[from] secp256k1::Error),
+
+    #[error("The transaction is partially signed")]
+    PartiallySigned,
+
+    #[error("The transaction is fully signed")]
+    FullySigned,
+}
+
+/// A wrapper enum that represents the transaction signed state. A transaction
+/// contained by this enum can be either fully signed or partially signed.
+pub enum Signed {
+    Fully(SignableTransaction),
+    Partially(SignableTransaction),
+}
+
+impl Signed {
+    /// Returns the transaction if it is fully signed, otherwise returns an error
+    pub fn fully_signed(self) -> std::result::Result<SignableTransaction, Error> {
+        match self {
+            Signed::Fully(tx) => Ok(tx),
+            Signed::Partially(_) => Err(Error::PartiallySigned),
+        }
+    }
+
+    /// Returns the transaction if it is partially signed, otherwise returns an error
+    pub fn partially_signed(self) -> std::result::Result<SignableTransaction, Error> {
+        match self {
+            Signed::Fully(_) => Err(Error::FullySigned),
+            Signed::Partially(tx) => Ok(tx),
+        }
+    }
+
+    /// Returns the transaction regardless of whether it is fully or partially signed
+    pub fn unwrap(self) -> SignableTransaction {
+        match self {
+            Signed::Fully(tx) => tx,
+            Signed::Partially(tx) => tx,
+        }
+    }
```

Comment:
Right...  I intentionally made it this way to prevent cloning...  and ironically this was done in response to feedback on the original implementation which was implementing  Ok(tx) Err(tx)..  I suppose we should perhaps add `try_fully_signed()` and such that can be recoverable...  I'll look at this in a bit.

Reply:
So I added `try_fully_signed()` and `try_partially_signed()` that route `tx` via `Ok(tx)` or `Err(tx)` depending on the state, allowing the recovery of the tx in either state.  Also, if a developer wants to manage state, he can do a direct match against the enum.

LMK if this looks ok.

Reply:
looks okay